### PR TITLE
[GPU] Optimize eltwise kernel for onednn formats

### PR DESCRIPTION
**Pull request by byungilm**
+ Bugfix of eltwise_b_fs_yx_fsv16 kernel for int satuation
+ Add optimizing for fsv16 using vload
+ Add optimizing for fsv32
+ Add test-cases to eltwise_gpu_test

Signed-off-by: Min, Byungil <byungil.min@intel.com>

### Details:
 - *item1*
 - *...*

### Tickets:
 - 74370

### CODE REVIEW

**Comment by byungilm** [comment]
list : failed TC
conv_fp32_add_per_element_planar_const
activation_eltwise_activation_quantize_u8
deconv_bias.basic/13
eltwise_test_mixed_layout.mixed_layout

File: src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+    k.EnableInputLayout(DataLayout::b_fs_zyx_fsv32);
+    k.EnableOutputLayout(DataLayout::b_fs_zyx_fsv32);
+    k.EnableInputLayout(DataLayout::bs_fs_yx_bsv32_fsv32);
+    k.EnableOutputLayout(DataLayout::bs_fs_yx_bsv32_fsv32);
+    k.EnableInputLayout(DataLayout::bs_fs_yx_bsv16_fsv32);
+    k.EnableOutputLayout(DataLayout::bs_fs_yx_bsv16_fsv32);
+    k.EnableInputLayout(DataLayout::bs_fs_yx_bsv32_fsv16);
+    k.EnableOutputLayout(DataLayout::bs_fs_yx_bsv32_fsv16);
+    k.EnableInputLayout(DataLayout::bs_fs_yx_bsv16_fsv16);
+    k.EnableOutputLayout(DataLayout::bs_fs_yx_bsv16_fsv16);
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp:43]: what about `zyx_bsv*_fsv*` family?

File: src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+        if ((GetFeatureBlockSizeFromFormat(ewParams, i) == 1) &&
+            !IsBroadcastingPossibleInput(ewParams.inputs[i], ewParams.outputs[0])) {
+            return false;
+        }
+    }
+
+    auto input0 = ewParams.inputs[0];
+    const auto vec_size = GetFeatureBlockSizeFromFormat(ewParams, 0);
+    const auto& output = ewParams.outputs[0];
+    // Check that padding before features doesn't miss-align the blocks
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp:104]: nit: `miss` -> `mis`

File: src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    if (input0.Feature().pad.before % vec_size != 0 || output.Feature().pad.before % vec_size != 0) {
+        return false;
+    }
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp:107]: Is it necessary condition? The spec does not say such restriction.. https://man.opencl.org/vloadn.html

File: src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+
+    auto compareTensors = [](const DataTensor& input0, const DataTensor& input1) -> bool {
+        // Check all parameters except DataType
+        auto& input0_dims = input0.GetDims();
+        auto& input1_dims = input1.GetDims();
+        bool same = input0.GetLayout() == input1.GetLayout() &&
+                    input0.GetPaddedVal() == input1.GetPaddedVal() &&
+                    input0.GetViewOffset() == input1.GetViewOffset() &&
+                    input0_dims.size() == input1_dims.size();
+        if (same) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp:117]: nit: I think this `if (same)` line can be removed.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+                        input0_dims[i].pad.before == input1_dims[i].pad.before &&
+                        input0_dims[i].pad.after == input1_dims[i].pad.after &&
+                        input0_dims[i].pitch == input1_dims[i].pitch;
+            }
+        }
+        return same;
+    };
+
+    for (size_t i = 1; i < ewParams.inputs.size(); i++) {
+        if (ewParams.inputs[i].LogicalSize() == input0.LogicalSize() && !(compareTensors(ewParams.inputs[i], input0)))
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp:129]: could you explain why we need to ensure tensors are the same? Is there a case where tensor shapes are different while the sizes are the same?

File: src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+            return false;
+        if (ewParams.inputs[i].Feature().pad.before % vec_size != 0) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp:131]: I think this is duplicated with line 105.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+    const auto &ew = params.operations[op_num];
+
+    const auto &input = ew.inputs[input_idx];
+    if (input.mode == EltwiseInputMode::INPUT_BUFFER) {
+        if (params.inputs[input_idx].LogicalSize() != 1
+            && params.inputs[input_idx].Feature().v == 1
+            && params.outputs[0].Feature().v != 1) {
+                return true;
+            }
+    }
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp:374]: this is basically same as `IsBroadCastingPossibleInput`. What about reusing the function?

**Review by isanghao — COMMENTED** [review]
Looks good in general

File: src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp (hunk 5 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp:129]: Isn't it possible to have same size with different DataLayout or with same size of different axes?

File: src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp (hunk 3 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp]: Error occur in crop_test/crop_gpu.pad_test
input format blocked format + planar format && feature size is not aligned

File: src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp (hunk 7 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp:380]: InputHasFeatureBroadcast and IsBroadCastingPossibleInput have a little bit different condition to return true. It needs additional condition to use IsBroadCastingPossibleInput here. Here the condition only finds an input which does not have vector size. But IsBroadCastingPossibleInput is different.
 So I think we do not need to reuse the code.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp (hunk 6 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/eltwise/eltwise_kernel_blocked_opt.cpp:137]: This also affects rop_test/crop_gpu.pad_test

**Comment by byungilm** [comment]
rebase done. ci/pot issue is cleared.

**Comment by isanghao** [comment]
Looks good to me. Let me check the performance.

**Review by isanghao — APPROVED** [review]

**Pull request merged**